libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -1759,6 +1782,101 @@ IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
+    uint8_t ** pPacketIdentifierHigh = NULL;
+    uint8_t * pMqttPacket = NULL;
+    size_t packetSize = 0;
+    uint16_t packetIdentifier;
+
+    IotMqtt_Assert( pPublishInfo != NULL );
+
+    /* Check whether the connection handle is valid. */
+    if( mqttConnection->disconnected == true )
+    {
+        status = IOT_MQTT_NETWORK_ERROR;
+    }
+
+    if( status == IOT_MQTT_SUCCESS )
+    {
+        /* Generate a PUBLISH packet from pPublishInfo. */
+        status = _getMqttPublishSerializer( mqttConnection->pSerializer )( pPublishInfo,
+                                                                           &( pMqttPacket ),
+                                                                           &( packetSize ),
+                                                                           &( packetIdentifier ),
+                                                                           pPacketIdentifierHigh );
+    }
+
+    if( status == IOT_MQTT_SUCCESS )
+    {
+        /* Transmit the MQTT packet from the operation over the network. */
+        if( mqttConnection->pNetworkInterface->send( mqttConnection->pNetworkConnection,
+                                                     pMqttPacket,
+                                                     packetSize ) != packetSize )
+        {
+            status = IOT_MQTT_NETWORK_ERROR;
+        }
+    }
+
+    /* Clean up memory for PUBLISH packet. */
+    if( pMqttPacket != NULL )
+    {
+        IotMqtt_FreeMessage( pMqttPacket );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
+                                     const IotMqttPublishInfo_t * pPublishInfo,
+                                     uint32_t flags,
+                                     const IotMqttCallbackInfo_t * pCallbackInfo,
+                                     IotMqttOperation_t * const pPublishOperation )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+
+    /* Check that IotMqtt_Init was called. */
+    if( _checkInit() == false )
+    {
+        status = IOT_MQTT_NOT_INITIALIZED;
+    }
+    /* Check whether the connection handle is valid. */
+    else if( mqttConnection->disconnected == true )
+    {
+        status = IOT_MQTT_NETWORK_ERROR;
+    }
+    else if( _IotMqtt_ValidatePublish( mqttConnection->awsIotMqttMode,
+                                       pPublishInfo,
+                                       flags,
+                                       pCallbackInfo,
+                                       pPublishOperation ) == false )
+    {
+        status = IOT_MQTT_BAD_PARAMETER;
+    }
+    else if( ( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
+               == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) &&
+             ( pPublishInfo->qos == IOT_MQTT_QOS_0 ) )
+    {
+        status = _publishQos0Sync( mqttConnection,
+                                   pPublishInfo );
```

Comment:
What is the callback behavior in this case? Are we completely ignoring the callback behavior? What will happen to applications already present and expecting a callback to mark the completion of publishAsync?

Reply:
As discussed offline, as this path is hit only for sync API, callbacks are not going to be present.

Reply:
We talked separately to clarify that a callback parameter is not expected for the sync Publish case as firstly, the `MQTT_INTERNAL_FLAG_BLOCK_ON_SEND` flag is not public-facing and thus, the customer cannot call the `IotMqtt_PublishAsync` function for a QoS 0 synchronous operation, and secondly, the sync Publish Qos 0 operation is requested by the `IotMqtt_PublishSync` function that passes the internal `MQTT_INTERNAL_FLAG_BLOCK_ON_SEND` flag to the `IotMqtt_PublishAsync` function.

Basically, the changes made do not change the behavior of the function (for any customer application depending on this function to be affected)!